pallets/gear/src/internal.rs

```diff
@@ -507,6 +507,92 @@ where
         mailboxed
     }
 
+    /// Delays dispatch sending.
+    pub(crate) fn send_delayed_dispatch(
+        origin_msg: MessageId,
+        dispatch: Dispatch,
+        delay: u32,
+        to_user: bool,
+    ) {
+        // Validating delay
+        if delay.is_zero() {
+            unreachable!("Delayed sending with zero delay appeared");
+        }
+
+        // Taking data for funds manipulations.
+        let from = <T::AccountId as Origin>::from_origin(dispatch.source().into_origin());
+        let value = dispatch.value().unique_saturated_into();
+
+        if to_user {
+            // Querying `MailboxThreshold`, that represents minimal amount of gas
+            // for message to be added to mailbox.
+            let threshold = T::MailboxThreshold::get();
+
+            // Figuring out gas limit for insertion.
+            //
+            // In case of sending with gas, we use applied gas limit, otherwise
+            // finding available funds and trying to take threshold from them.
+            let gas_limit = dispatch
+                .gas_limit()
+                .or_else(|| {
+                    // Querying gas limit. Fails in cases of `GasTree` invalidations.
+                    let gas_limit = GasHandlerOf::<T>::get_limit(origin_msg)
+                        .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+
+                    // If available gas is greater then threshold,
+                    // than threshold can be used.
+                    (gas_limit >= threshold).then_some(threshold)
+                })
+                .unwrap_or_default();
+
+            if !dispatch.is_error_reply() && gas_limit >= threshold {
```

Comment:
what if `gas_limit` is zero (in case `origin_msg` has less gas than `threshold`)? Seems no cut happens, but task for the message wake is added. Is it correct?

Reply:
Also, is error reply supposed to be added to the task pool?

Reply:
Does message delay logic require paying for being delayed, i.e. being in the `TaskPool`?

Reply:
discussed